py/qstr.c

```diff
@@ -171,14 +194,48 @@ STATIC qstr qstr_add(const byte *q_ptr) {
     return MP_STATE_VM(last_pool)->total_prev_len + MP_STATE_VM(last_pool)->len - 1;
 }
 
+#define MP_QSTR_SEARCH_THRESHOLD 10
+
 qstr qstr_find_strn(const char *str, size_t str_len) {
-    // work out hash of str
     mp_uint_t str_hash = qstr_compute_hash((const byte *)str, str_len);
 
     // search pools for the data
     for (qstr_pool_t *pool = MP_STATE_VM(last_pool); pool != NULL; pool = pool->prev) {
-        for (const byte **q = pool->qstrs, **q_top = pool->qstrs + pool->len; q < q_top; q++) {
-            if (Q_GET_HASH(*q) == str_hash && Q_GET_LENGTH(*q) == str_len && memcmp(Q_GET_DATA(*q), str, str_len) == 0) {
+        size_t low = 0;
+        size_t high = pool->len - 1;
+
+        // binary search inside the pool
+        if (pool->sorted) {
+            while (high - low > MP_QSTR_SEARCH_THRESHOLD) {
+                size_t mid = (low + high + 1) / 2;
+                const byte **q = pool->qstrs + mid;
+                size_t len = Q_GET_LENGTH(*q);
+                if (len > str_len) {
+                    len = str_len;
+                }
+                int cmp = memcmp(Q_GET_DATA(*q), str, str_len);
+                if (cmp < 0) {
+                    low = mid;
+                } else if (cmp > 0) {
+                    high = mid;
+                } else {
+                    if (Q_GET_LENGTH(*q) < str_len) {
+                        low = mid;
+                    } else if (Q_GET_LENGTH(*q) > str_len) {
+                        high = mid;
+                    } else {
+                        return pool->total_prev_len + (q - pool->qstrs);
+                    }
+                }
+            }
+        }
+
+        // sequential search for the remaining strings
+        for (const byte **q = pool->qstrs + low; q != pool->qstrs + high + 1; q++) {
+            if (*q &&
```

Comment:
@amirgon why does it check for `*q` being non-zero?  The original code did not need to do this in the linear search part.

Reply:
I removed this check after rebase.